esmvalcore/preprocessor/_io.py

```diff
@@ -98,65 +108,63 @@ def load(
     ------
     ValueError
         Cubes are empty.
+    TypeError
+        Invalid type for ``file``.
+
     """
-    file = Path(file)
-    logger.debug("Loading:\n%s", file)
+    if isinstance(file, (str, Path)):
+        cubes = _load_from_file(file, ignore_warnings=ignore_warnings)
+    elif isinstance(file, Cube):
+        cubes = CubeList([file])
+    elif isinstance(file, CubeList):
+        cubes = file
+    elif isinstance(file, (xr.Dataset, ncdata.NcData)):
+        cubes = dataset_to_iris(file, ignore_warnings=ignore_warnings)
+    else:
+        raise TypeError(
+            f"Expected type str, pathlib.Path, iris.cube.Cube, "
+            f"iris.cube.CubeList, xarray.Dataset, or ncdata.NcData for file, "
+            f"got type {type(file)}"
+        )
 
-    if ignore_warnings is None:
-        ignore_warnings = []
+    if not cubes:
+        raise ValueError(f"{file} does not contain any data")
 
-    # Avoid duplication of ignored warnings when load() is called more often
-    # than once
-    ignore_warnings = list(ignore_warnings)
+    for cube in cubes:
+        if "source_file" not in cube.attributes:
+            logger.warning(
```

Comment:
Should we raise an exception for this? If it is only expected to happen when people contribute new fixes, I would raise an exception to make sure we don't bother users with many warnings. On the other hand, if it could happen to users, maybe it should be a debug log message instead?

Reply:
It might be better to raise an actual `ESMValCoreUserWarning` instead of using the logger since this problem can only occur when using the Python API. This makes it much more straightforward to ignore the warning if required. @bouweandela what do you think?